multimap/src/main/java/org/infinispan/multimap/impl/HashMapBucket.java

```diff
@@ -0,0 +1,130 @@
+package org.infinispan.multimap.impl;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import org.infinispan.commons.marshall.ProtoStreamTypeIds;
+import org.infinispan.marshall.protostream.impl.MarshallableUserObject;
+import org.infinispan.multimap.internal.MultimapDataConverter;
+import org.infinispan.protostream.annotations.ProtoFactory;
+import org.infinispan.protostream.annotations.ProtoField;
+import org.infinispan.protostream.annotations.ProtoTypeId;
+
+@ProtoTypeId(ProtoStreamTypeIds.MULTIMAP_HASH_MAP_BUCKET)
+public class HashMapBucket<K, V> {
+
+   final Map<Object, Object> values;
+
+   // We don't transfer the converter in the proto message.
+   private MultimapDataConverter<K, V> converter = null;
```

Comment:
I am not a fan of the converter being stored in the HashMapBucket itself.. it can just be an argument that is used for the putAll method and create methods. It is a request variable not a state variable.

Reply:
Updated.